src/main/java/org/datadog/jmxfetch/App.java

```diff
@@ -472,14 +472,15 @@ void start() {
             try {
                 long loopPeriod = appConfig.getCheckPeriod();
                 long sleepPeriod = loopPeriod - duration;
-                if (sleepPeriod <= 0) {
+                if (sleepPeriod < loopPeriod / 2) {
                     log.debug(
                         "The collection cycle took longer that the configured check period,"
-                        + " the next cycle will start immediatly");
+                        + " the next cycle will be delayed");
+                    sleepPeriod = loopPeriod / 2;
                 } else {
                     log.debug("Sleeping for " + sleepPeriod + " ms.");
-                    Thread.sleep(sleepPeriod);    
                 }
+                Thread.sleep(sleepPeriod);
```

Comment:
I think Thread.sleep() for a negative value will throw a `java.lang.IllegalArgumentException`. Please bear in mind that some instances might take longer to collect than the actual loopPeriod. That would result in a negative sleepPeriod that could trigger that exception.  

The goal of this part of the code was to ensure we didn't skip iterations (and sleep) when the duration of the collection exceeded the check period, I think this change could change that expected behavior.

Maybe I'm missing something, feel free to ping me (or drop it in here) if you want to provide more context or discuss.

Reply:
@truthbk 

> I think Thread.sleep() for a negative value will throw a java.lang.IllegalArgumentException. Please bear in mind that some instances might take longer to collect than the actual loopPeriod. That would result in a negative sleepPeriod that could trigger that exception.

`sleepPeriod` will not be negative, because it is assigned a new value a few lines above.

> The goal of this part of the code was to ensure we didn't skip iterations (and sleep) when the duration of the collection exceeded the check period, I think this change could change that expected behavior.

True, but when the slowness is caused by the bean update process, two metric collections happen within a short time, sending two sets of metrics within one period. This is useless for gauges, skews histograms a bit and can cause confusing spikes in reported counter rates (small change over tiny time interval yields very high rate). Given that this behavior is actually relatively new, we might take a step back until a better solution is worked out. WDYT?